### PR TITLE
RFC 1008: TLS and ALPN Support

### PR DESCRIPTION
**Pull request by fantix**
[rendered](https://github.com/fantix/rfcs/blob/tls-v2/text/1008-tls-and-alpn.rst)

Closes #36

### CODE REVIEW

File: text/1008-tls-and-alpn.rst
Comment (fmoor) [inline text/1008-tls-and-alpn.rst:298]: Does this mean that we would undeprecate the currently deprecated admin unix socket features?

Comment (tailhook) [reply text/1008-tls-and-alpn.rst:298]: I don't think so. It's still an appropriate way to use for auxiliary needs like resetting password, but I don't think we need it for anything else.

E.g. I don't think this is a good way for batch processing of data as disk usage time will probably dominate on the encryption time anyway.

File: text/1008-tls-and-alpn.rst
Comment (tailhook) [inline text/1008-tls-and-alpn.rst:319]: We don't need storing client certificate fingerprints.

The idea of client certificates is:
1. We have a CA, certificate of which is know to the EdgeDB server
2. CA emits a client certificate, having some identifier of the user name, say common-name of `username.yourdb.example.org`
3. When client authenticates it sends the certificate from (2) to the EdgeDB server, server verifies it via certificate described in (1) [this involves client-certificate private key to encode some challenge-responses but I'm omitting the details now].

So on each authorization, client certificate is checked against certificate chain and no individual user info is stored between sesesions. (Well, there is TSL session cache, but that's a different layer of abstraction).

File: text/1008-tls-and-alpn.rst
Comment (tailhook) [inline text/1008-tls-and-alpn.rst:357]: Can we have `--allow-non-tls` rather than `--no-tls` ? I think we should be able to detect whether the handshake was TLS one or binary protocol.

File: text/1008-tls-and-alpn.rst
Comment (tailhook) [inline text/1008-tls-and-alpn.rst:386]: This will not be the case if my proposal above is accepted.

File: text/1008-tls-and-alpn.rst
Comment (tailhook) [inline text/1008-tls-and-alpn.rst:375]: You don't need CA-based trust to prevent MITM. Veryfying server certificate is enough. So generating `credentials.json` by web UI of cloud (Aether) or providing certificate fingerprint alongside of user and password from DB Ops to Users within the company is enough for that. Client certificates are even better.

File: text/1008-tls-and-alpn.rst
Comment (elprans) [inline text/1008-tls-and-alpn.rst:163]: I'd clarify that `tls_certdata` can specify an issuing certificate as well up to some long-lived root cert.

File: text/1008-tls-and-alpn.rst
Comment (elprans) [inline text/1008-tls-and-alpn.rst:254]: Specifically, this would be `http/1.1`, as we might also add support for `h2` in the future too.

File: text/1008-tls-and-alpn.rst
Comment (fantix) [reply text/1008-tls-and-alpn.rst]: Yeah sure thing - a real compatible dev server could be more handy.

File: text/1008-tls-and-alpn.rst
Comment (fantix) [reply text/1008-tls-and-alpn.rst]: Let's skip hostname check for now, and the client could have a `TLS_STRICT` mode in the future if we need the check.

File: text/1008-tls-and-alpn.rst
Comment (fantix) [reply text/1008-tls-and-alpn.rst]: Thanks for the comments and the offline discussion! To summarize, we'll move the server parameters to `metadata.json`, and leave the `credentials.json` for only client connection parameters (public key/cert only). We're not introducing CA just yet - this could come together with the future client cert verification.

File: text/1008-tls-and-alpn.rst
Comment (fantix) [reply text/1008-tls-and-alpn.rst]: I've updated the RFC not to use ``credentials.json`` for private keys. But I don't think we can use the CLI-specific data dir for core EdgeDB development, and the dev server doesn't always have a stable data dir (e.g. with `--temp-dir` or `--postgres-dsn`). So I kept some of the previous design (CLI generates a new self-signed cert and send to server through stdout on startup), and updated the pub cert into ``credentials.json`` together with the new ``edgedb authenticate`` command for easier core dev.

File: text/1008-tls-and-alpn.rst
Comment (fantix) [reply text/1008-tls-and-alpn.rst:163]: 👍 Thanks!  Added an open question - looking forward to more comments on this topic.

File: text/1008-tls-and-alpn.rst
Comment (1st1) [inline text/1008-tls-and-alpn.rst:80]:
```suggestion
(This is exactly ``ssl.SSLContext.load_cert_chain()`` from Python [4]_.)
```

Single backticks mean something else in ReST, please check if you use them elsewhere

File: text/1008-tls-and-alpn.rst
Comment (1st1) [inline text/1008-tls-and-alpn.rst:138]:
```suggestion
Likewise, the ``metadata.json`` file will be updated with the full path
```

File: text/1008-tls-and-alpn.rst
Comment (1st1) [reply text/1008-tls-and-alpn.rst:378]: Yes, we should clarify that the support of admin UNIX socket will be left only in the CLI. All bindings will not be able to connect to UNIX sockets at all.

**Review by 1st1 — APPROVED** [review]
LGTM. I think we should merge this and iterate in follow up PRs if necessary (I think there's only one or two yet unaddressed comments)

File: text/1008-tls-and-alpn.rst
Comment (elprans) [inline text/1008-tls-and-alpn.rst:294]: TLS <1.2 isn't considered to be secure anymore, so I think we should have 1.2 as the minimum.

**Pull request merged**
